parser: change SESSION USER to use session_user

otan opened the pull request:
Correct SESSION USER to actually use the session_user builtin. This
should not affect default expressions because they store the full
builtin name instead of the parsing syntax.


Resolves #69268

Release note (backward-incompatible change, sql change): Using `SESSION
USER` in a projection or where clause now actually returns the SESSION
USER instead of the CURRENT USER. If you require this to be backwards
compatible, use `session_user()` for SESSION USER and `current_user()`
or `CURRENT USER` if you need the CURRENT USER.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/70444)
<!-- Reviewable:end -->

rafiss commented:
Nice, I definitely feel like we should fix, but this is another example of how we need a better story for how we decide which "breaking" bug fixes to make. I'll refer to this as an example when asking other leads about breaking changes.

otan commented:
yeah at least with this one we want the backwards incompatible route, but it sucks for 21.2 users

otan committed:
parser: change SESSION USER to use session_user

Correct SESSION USER to actually use the session_user builtin. This
should not affect default expressions because they store the full
builtin name instead of the parsing syntax.

Release note (backward-incompatible change, sql change): Using `SESSION
USER` in a projection or where clause now actually returns the SESSION
USER instead of the CURRENT USER. If you require this to be backwards
compatible, use `session_user()` for SESSION USER and `current_user()`
or `CURRENT USER` if you need the CURRENT USER.

otan commented:
on second thought, i think we can backport this to v21.1. This is because during that mixed state, `SESSION USER` will evaluate at the parser to `current_user()`, and in v21.1 to `session_user()`.

rafiss reviewed (approved):
this is def the right thing to do now.

though just a heads up to you, as part of https://github.com/cockroachdb/cockroach/issues/54696#issuecomment-885814808 @RichardJCai will be changing CURRENT_USER and SESSION_USER so they don't parse as  functions

rafiss commented:
makes sense to backport to 21.2.1 i think

otan commented:
bors r=rafiss
